### ISSUE TITLE
Issue with reserved keyword reserving wrong fields

### ISSUE THREAD

**Issue by Ycallaer** [protocolbuffers/protobuf#9854]
Hi all,
Hope you can help a bit with below situation
Uploaded files:
[generic_mandate_reference_test.proto.txt](https://github.com/protocolbuffers/protobuf/files/8562878/generic_mandate_reference_test.proto.txt)
[generic_mandate_reference_test_descriptor.proto.txt](https://github.com/protocolbuffers/protobuf/files/8562879/generic_mandate_reference_test_descriptor.proto.txt)

Info on the situation:

**What version of protobuf and what language are you using?**
Version: 3.20.1
OS: Linux (tried both ubuntu and alpine)

**What runtime / compiler are you using (e.g., python version or gcc version)**
bash-5.1# gcc --version
gcc (Alpine 10.2.1_pre1) 10.2.1 20201203

**What did you do?**
1. We have a proto file (generic_mandate_reference_test.proto.txt, just strip txt off) that has the following definition
```
syntax = "proto3";

package iit.datahub.payments.v1;

// Generic mandate reference that is used by multiple events
message GenericMandateReference {

    // Mandate identifier
    string mandate_id = 1;

    // External mandate identifier used by end-clients, saxo, aggregators and correpondent banks to ascertain a particular mandate
    string external_mandate_id = 2;

    // Client identifier
    string client_id = 3;

     reserved 4;
     reserved "status_details";

     // Direct Debit Authorization mandate state
    string mandate_state = 5;

    // Direct Debit status details
    repeated string direct_debit_status_details = 6;
}
```

If we use protoc to generate a descriptor file, the reserved mapping is incorrectly parsed.
Command we run in the root of the proto directory:
```
for p in $(find proto/ -name '*.proto')
do
	protoc --descriptor_set_out=${p%.proto}.pd --proto_path=proto --proto_path=/tmp/external/proto --proto_path=/tmp/external/common "$p"
done
```

**What did you expect to see**

```
syntax = "proto3";
package iit.datahub.payments.v1;

message GenericMandateReference {
  reserved 4 ;
  reserved "status_details";

  string mandate_id = 1 [json_name = "mandateId"];
  string external_mandate_id = 2 [json_name = "externalMandateId"];
  string client_id = 3 [json_name = "clientId"];
  string mandate_state = 5 [json_name = "mandateState"];
  repeated string direct_debit_status_details = 6 [json_name = "directDebitStatusDetails"];
}
```

**What did you see instead?**

```
syntax = "proto3";
package iit.datahub.payments.v1;

message GenericMandateReference {
  reserved 4 to 5;
  reserved "status_details";

  string mandate_id = 1 [json_name = "mandateId"];
  string external_mandate_id = 2 [json_name = "externalMandateId"];
  string client_id = 3 [json_name = "clientId"];
  string mandate_state = 5 [json_name = "mandateState"];
  repeated string direct_debit_status_details = 6 [json_name = "directDebitStatusDetails"];
}
```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment**
We are currently on 3.17.3 but upgrading to latest version did not fix it.
Interpreting the pd file from protoc with either Java or Python3.6 shows the same result, the reserved keyword is mapping wrongly

**Comment by acozzette** [comment]
How did you convert the serialized FileDescriptorSet back to a proto file? The reason I ask is that in the descriptor representation, the end of the range is exclusive and so we would expect to see a 5 there in the serialized descriptor: https://github.com/protocolbuffers/protobuf/blob/6718fa9dcc7c5801cf65221317e3faa4a50f48ee/src/google/protobuf/descriptor.proto#L120

**Comment by Ycallaer** [comment]
So if I use python, I use the following piece of code
```
from google.protobuf.descriptor_pb2 import FileDescriptorSet

with open('/home/yves/generic_mandate_reference.pd', mode='rb') as f:
  data = f.read()

fd = FileDescriptorSet.FromString(data)
print(fd)
# Try to print out the options but will not print anything
print("Reserved: " + str(fd.file[0].message_type[0].reserved_range))
```

From which I can see the following partial output:
```
reserved_range {
      start: 4
      end: 5
    }
```

However our use case is in Java, where we take the file and cast it to a Confluent schema as follows

```
public static void buildProtofromDescriptor(String arg) throws IOException {
        //Read a descriptor file into a FileDescriptorSet
        DescriptorProtos.FileDescriptorSet dp =
                DescriptorProtos.FileDescriptorSet
                        .parseFrom(Files
                                .readAllBytes(
                                        Paths.get(arg)));

        String protoc_encoded = Base64.getEncoder().encodeToString(dp.getFile(0).toByteArray());

        // Then the same as before, as provided by Confluent:
        ProtobufSchema ps = new ProtobufSchema(protoc_encoded);
        File pd = new File(arg);
        String fileName = pd.getName().replaceAll("\\.pd$", ".proto");
        BufferedWriter writer = new BufferedWriter(new FileWriter(pd.getParentFile() + "/" + fileName));
        writer.write(ps.canonicalString());
        writer.close();
```

In both cases it generates a protobuf schema with a range reservation 4 to 5.

**Comment by Ycallaer** [comment]
So I tried something this morning. I upgraded the Confluent library to the latest version and now it parses correctly.
It was insightful to learn that the range is exclusive, it seems that was an implementation bug in Confluent kafka-protobuf-serializer library.
I upgraded to 7.1.0 and now I get the following output
```
message GenericMandateReference {
  reserved 4;
  reserved "status_details";
```
which is what I expected. So this issue can be closed, but much appreciated for the help and insights.
Thanks.
